### PR TITLE
Adding a show_read override to GetPosts.

### LINKED ISSUE #4124: Ignore user filters per request
### Requirements

- [X] Is this a feature request? For questions or discussions use https://lemmy.ml/c/lemmy_support
- [X] Did you check to see if this issue already exists?
- [X] Is this only a feature request? Do not put multiple feature requests in one issue.
- [X] Is this a backend issue? Use the [lemmy-ui](https://github.com/LemmyNet/lemmy-ui) repo for UI / frontend issues.

### Is your proposal related to a problem?

- Apps may have their own implementations of "hide read posts" (for example, Voyager). These client side implementations may conflict with lemmy's implementation baked into the API based on user settings, confusing users
- Apps may want to retrieve all posts ignoring any backend filters such as "hide read posts" setting for arbitrary requests: for example, to fetch posts in moderated sub, on users' profiles, for special apps, integrations and filters, etc.

### Describe the solution you'd like.

Every endpoint where posts are retrieved, for example, `getPosts`, `getPersonDetails` etc, has a request parameter `show`.

 - (not present on request) - use user default, hide posts according to `show_read_posts` user setting etc - **maintain current behavior**
 - `show: "all"` Return all posts. Don't hide posts based on user settings.

This would be a backward compatible change. It would also be forward compatible, in case Lemmy adds additional backend user filters, for example "hide posts I've voted on".

### Describe alternatives you've considered.

afaik there's not currently a good way around this.

- You could fetch posts without being authenticated, but then you don't get `my_vote`, etc.
- You could quickly turn the user setting off, make request for posts, then turn back on, but that's not ideal for obvious reasons
- Apps could message user to turn the setting off for their account. But, they might prefer it to be on for using other clients like lemmy-ui.

### Additional context

Reddit has something similar to overcome this problem-

https://www.reddit.com/dev/api/#listings

> show - optional parameter; if all is passed, filters such as "hide links that I have voted on" will be disabled.

### PR DESCRIPTION
**Pull request by dessalines**
- If show_read is true, it overrides the local user show_read setting.
- Fixes #4124

### CODE REVIEW

**Review by SleeplessOne1917 — APPROVED** [review]

**Comment by dullbananas** [comment]
`show_read: Some(false)` doesn't override the user setting, which could cause confusion. Use `Option<bool>` in `PostQuery`.

**Comment by dessalines** [comment]
I've updated to make all of the post and comment query params as options.

**Review by dullbananas — CHANGES_REQUESTED** [review]
Changing to `Option<bool>` is not needed for fields other than `show_read` because they have the same default value every time, unlike `show_read` which overrides a `LocalUser` field.

If the inconsistency of using both `bool` and `Option<bool>` is a problem, then you can either:
* Keep these changes (except for the change to `page_back` in my opinion)
* Use `bool` instead of `Option<bool>`, remove the `Default` derive, and create this function that's does the same thing as `Default::default` except it sets `local_user: local_user, show_read: local_user.show_read_posts()`:
  ```
  fn new(local_user: Option<&'a LocalUser>) -> Self
  ```

File: crates/db_views/src/post_view.rs [hunk 1]
```diff
@@ -400,22 +400,22 @@ fn queries<'a>() -> Queries<
...
-    if options.saved_only {
+    if options.saved_only.unwrap_or_default() {
       query = query
         .filter(post_saved::person_id.is_not_null())
         .then_order_by(post_saved::published.desc());
     }
     // Only hide the read posts, if the saved_only is false. Otherwise ppl with the hide_read
     // setting wont be able to see saved posts.
-    else if !options.local_user.show_read_posts() {
+    else if !(options.show_read.unwrap_or_default() || options.local_user.show_read_posts()) {
```
[the remark this answers is not in the document]
Comment (dessalines) [reply crates/db_views/src/post_view.rs]: Thx, I've added this one, but I think I would like to keep these as Options.

**Review by Nutomic — APPROVED** [review]

**Review by dullbananas — APPROVED** [review]

**Pull request merged**